### ISSUE TITLE
Getting data from redis getting slowly when redis-server disconnect

### ISSUE THREAD

**Issue by yangjiaronga** [Grokzen/redis-py-cluster#333]
Redis-server disconnect when Redis-client not send any data during 300s
---

Redis-py-cluster get data raise ConnectionError, setting try_random_node is True.

continue while, get node is possibly not key slot node, raise MovedError,
setting refresh_table_asap = True.

next, get data to need refresh table asap time-consuming and 
node-connect the only release is not disconnected

I think that when raising ConnectionError:
1. disconnect
2. get other connection from the current node
3. if rasing ConnectionError when getting connection is new connection(make_connection), else ConnectionError

```python
@clusterdown_wrapper
def execute_command(self, *args, **kwargs):
    """
    Send a command to a node in the cluster
    """
    if not args:
        raise RedisClusterException("Unable to determine command to use")

    command = args[0]

    # If set externally we must update it before calling any commands
    if self.refresh_table_asap:
        self.connection_pool.nodes.initialize()
        self.refresh_table_asap = False

    node = self.determine_node(*args, **kwargs)
    if node:
        return self._execute_command_on_nodes(node, *args, **kwargs)

    redirect_addr = None
    asking = False

    try_random_node = False
    slot = self._determine_slot(*args)
    ttl = int(self.RedisClusterRequestTTL)

    while ttl > 0:
        ttl -= 1

        if asking:
            node = self.connection_pool.nodes.nodes[redirect_addr]
            r = self.connection_pool.get_connection_by_node(node)
        elif try_random_node:
            r = self.connection_pool.get_random_connection()
            try_random_node = False
        else:
            if self.refresh_table_asap:
                # MOVED
                node = self.connection_pool.get_master_node_by_slot(slot)
            else:
                node = self.connection_pool.get_node_by_slot(slot)
            r = self.connection_pool.get_connection_by_node(node)

        try:
            if asking:
                r.send_command('ASKING')
                self.parse_response(r, "ASKING", **kwargs)
                asking = False

            r.send_command(*args)
            return self.parse_response(r, command, **kwargs)
        except (RedisClusterException, BusyLoadingError):
            raise
        except (ConnectionError, TimeoutError):
            try_random_node = True

            if ttl < self.RedisClusterRequestTTL / 2:
                time.sleep(0.1)
        except ClusterDownError as e:
            self.connection_pool.disconnect()
            self.connection_pool.reset()
            self.refresh_table_asap = True

            raise e
        except MovedError as e:
            # Reinitialize on ever x number of MovedError.
            # This counter will increase faster when the same client object
            # is shared between multiple threads. To reduce the frequency you
            # can set the variable 'reinitialize_steps' in the constructor.
            self.refresh_table_asap = True
            self.connection_pool.nodes.increment_reinitialize_counter()

            node = self.connection_pool.nodes.set_node(e.host, e.port, server_type='master')
            self.connection_pool.nodes.slots[e.slot_id][0] = node
        except TryAgainError as e:
            if ttl < self.RedisClusterRequestTTL / 2:
                time.sleep(0.05)
        except AskError as e:
            redirect_addr, asking = "{0}:{1}".format(e.host, e.port), True
        finally:
            self.connection_pool.release(r)

    raise ClusterError('TTL exhausted.')
```

**Comment by Grokzen** [comment]
@yangjiaronga I do not understand your bug report.  You need to re-format this issue with more exact instructions on how to reproduce this error you are seeing becuase you are just quoting most of the code without any steps for me to reproduce the error. I need to see your client code and what scenario of the cluster you are using/building etc.

**Comment by yangjiaronga** [comment]
Thanks a lot for the reply.

Redis-server config
```redis
127.0.0.1:6379> config get timeout
1) "timeout"
2) "300"
```
If there are more than 300 seconds between two requests, a second request is triggered ConnectionError

I do not understand setting try_random_node  = True when raising  ConnectionError
```python
except (ConnectionError, TimeoutError):
    try_random_node = True

    if ttl < self.RedisClusterRequestTTL / 2:
        time.sleep(0.1)
except ClusterDownError as e:
    self.connection_pool.disconnect()
    self.connection_pool.reset()
    self.refresh_table_asap = True

    raise e
```

I think the code should look like this
```python
except ConnectionError:
    # try_random_node = True
    r.disconnect()
except TimeoutError:
    # try_random_node = True
    if ttl < self.RedisClusterRequestTTL / 2:
        time.sleep(0.1)
```

because set try_random_node = True，exclude r = self.connection_pool.get_random_connection()
Might trigger MovedError

```python
if asking:
    node = self.connection_pool.nodes.nodes[redirect_addr]
    r = self.connection_pool.get_connection_by_node(node)
elif try_random_node:
    r = self.connection_pool.get_random_connection()
    try_random_node = False
```

finally, I don't want a ConnectionError to cause a MovedError

**Comment by Grokzen** [comment]
One other ting. You say that you do not want a `MovedError` to be triggered on a ConnectionError pathway, can you shoow with a example where you would get a MovedError raised up to your side of the code? or would this only be internally within the client lib that a MovedError would be raised? becuase most likley that MovedError would recorrect itself anyway in the next loop run of the main attempt loop.

**Comment by yangjiaronga** [comment]
I agree with trying a random node that this case and exception kinda assumes that whats happen is that one node just fallen off for some reason and that your cluster basically lost one node.

There are two possible reasons for triggering a connection error, lost one node(master node down) and node initiative disconnect.  But you've unified the two cases, I think it's wrong.

in my test script:
if run sleep(1) and then print(rc.get('foo')) again and time that call,  i get 0.001 time response.

I have no problem closing the connection when there is a connection error, But a TTL error is thrown, we be thinking about how to solve the problem of TTL error, I think you can avoid TTL errors by handling the two cases of connection errors

this only be internally within the client lib that a MovedError would be raised

This is how I found the problem
We switched from redis singleton to redis cluster in production environment
Peak redis cluster performance is 10% slower than redis single instance performance，about qps = 5W/s
Redis clusters are 5-10 times slower than redis singletons during low peaks, about qps = 30/s，Sometimes very quickly

Then I looked at the redis cluster source code
I found the reason for the slowness，Whenever running this code, It takes between 50 and 300 milliseconds
```python
if self.refresh_table_asap:
    self.connection_pool.nodes.initialize()
    self.refresh_table_asap = False
```

And then I look up where setting self.refresh_table_asap = True
ClusterDownError and MovedError
I am sure my cluster is not down and no exceptions were caught, so is MovedError
```python
except ClusterDownError as e:
    self.connection_pool.disconnect()
    self.connection_pool.reset()
    self.refresh_table_asap = True
    raise e
except MovedError as e:
    self.refresh_table_asap = True
    self.connection_pool.nodes.increment_reinitialize_counter()
    node = self.connection_pool.nodes.set_node(e.host, e.port, server_type='master')
    self.connection_pool.nodes.slots[e.slot_id][0] = node
```

So what do I find that causes a MovedError
This code can get the correct node, `self.connection_pool.get_node_by_slot(slot)`

This code will get a random node, which may trigger a MovedError `self.connection_pool.get_random_connection()`

Because I found the ConnectionError, because it will set try_random_node = True

```redis
127.0.0.1:6379> config get timeout
1) "timeout"
2) "300"
```
Too few requests during the low peak period will result in no requests within 300s, leading to active disconnection of redis-server

**Comment by Grokzen** [comment]
@yangjiaronga I think this has now been fixed in the following commit c62012c7bfaf3ffe4f199de39085fe355fcc5737 where the first proposed change was implemented. If it is possible, could you try the latest version of master or a commit that is after the mentioned commit to see if you still see the pattern or the problem now after this?

**Comment by Grokzen** [comment]
@yangjiaronga I did attempt to run your example script that has the timeout set on the node to 20 seconds in my case and this is the following result that i got from running the script

```
(redis) ➜  redis-py-cluster git:(master) ✗ python a.py
0.00032639503479003906                                
bar                                                   
0.002065896987915039                                  
```

Yes there is a slowdown when running the second pass of x10, but that is more in line of an acceptable performance compared to the numbers you had where you had almost an x1000 increase in latency/lagg when the cluster was disconnected.

Right now i will consider this issue resolved after the tests i have done for now. If you see that this error still remains in some way/shape/form then please either comment here and we can discuss it (please do not reopen it), or please open a new issue if there is other details or other scenarios where this happens.
